internal/pipe/ko/ko.go

```diff
@@ -0,0 +1,269 @@
+		if err != nil {
+			return fmt.Errorf("NewGo: %v", err)
+		}
+		r, err := b.Build(ctxBackground, opts.ip)
+		if err != nil {
+			return fmt.Errorf("build: %v", err)
+		}
+
+		namer := options.MakeNamer(&options.PublishOptions{
+			DockerRepo:          opts.dockerRepo,
+			Bare:                opts.bare,
+			PreserveImportPaths: opts.preserverImportPaths,
+			BaseImportPaths:     opts.baseImportPaths,
+		})
+
+		p, err := publish.NewDefault(opts.dockerRepo,
+			publish.WithTags(opts.tags),
+			publish.WithNamer(namer),
+			publish.WithAuthFromKeychain(authn.DefaultKeychain))
+		if err != nil {
+			return fmt.Errorf("NewDefault: %v", err)
+		}
+		ref, err := p.Publish(ctxBackground, r, opts.ip)
+		if err != nil {
+			return fmt.Errorf("publish: %v", err)
+		}
+		fmt.Println(ref.String())
+		return nil
+	}
+}
+
+func fromConfig(ctx *gcontext.Context, cfg config.Ko) (*buildOptions, error) {
+	if cfg.WorkingDir == "" {
+		wd, err := os.Getwd()
+		if err != nil {
+			return nil, err
+		}
+
+		cfg.WorkingDir = wd
+	}
+
+	localImportPath := fmt.Sprint(".", string(filepath.Separator), ".")
+
+	dir := filepath.Clean(cfg.WorkingDir)
+	if dir == "." {
+		dir = ""
+	}
+
+	pkgs, err := packages.Load(&packages.Config{Mode: packages.NeedName, Dir: dir}, localImportPath)
+	if err != nil {
+		return nil, fmt.Errorf("'builds': %s does not contain a valid local import path (%s) for directory (%s): %w", cfg.ID, localImportPath, cfg.WorkingDir, err)
+	}
+
+	if len(pkgs) != 1 {
+		return nil, fmt.Errorf("'builds': %s results in %d local packages, only 1 is expected", cfg.ID, len(pkgs))
+	}
+
+	opts := &buildOptions{
+		ip:                   pkgs[0].PkgPath,
+		workingDir:           cfg.WorkingDir,
+		bare:                 cfg.Bare,
+		preserverImportPaths: cfg.PreserveImportPaths,
+		baseImportPaths:      cfg.BaseImportPaths,
+	}
+
+	if cfg.BaseImage != "" {
+		opts.baseImage = cfg.BaseImage
+	} else {
+		opts.baseImage = "cgr.dev/chainguard/static" // TODO: we can discuss on this
+	}
+
+	if cfg.Platforms != nil {
+		opts.platforms = cfg.Platforms
+	} else {
+		opts.platforms = []string{"linux/amd64"}
+	}
+
+	if cfg.Tags != nil {
+		opts.tags = cfg.Tags
+	} else {
+		opts.tags = []string{"latest"}
+	}
+
+	if cfg.SBOM != "" {
+		opts.sbom = cfg.SBOM
+	} else {
+		opts.sbom = "spdx"
+	}
+
+	if ctx.Env["KO_DOCKER_REPO"] != "" {
+		opts.dockerRepo = ctx.Env["KO_DOCKER_REPO"]
+	} else {
+		opts.dockerRepo = cfg.Repository
+	}
+
+	if ctx.Env["COSIGN_REPOSITORY"] != "" {
+		opts.cosignRepo = ctx.Env["COSIGN_REPOSITORY"]
+	} else {
+		opts.cosignRepo = cfg.CosignRepository
+	}
+
+	cfg.LDFlags = []string{}
+
+	var ldflags []string
+	if len(cfg.LDFlags) != 0 {
+		if cfg.Pkg == "" {
+			return nil, fmt.Errorf("'builds': %s requires a package to be specified", cfg.ID)
+		}
+		for _, lf := range cfg.LDFlags {
+			tlf, err := tmpl.New(ctx).Apply(lf)
+			if err != nil {
+				return nil, err
+			}
+			ldflags = append(ldflags, tlf)
+		}
+
+		opts.ldflags = ldflags
+	}
```

Comment:
i think all or most of this can be in `Default`

Reply:
yes